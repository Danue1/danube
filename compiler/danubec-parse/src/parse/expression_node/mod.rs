mod literal_expression_node;

use crate::State;

impl crate::Context {
    pub fn expression_node(&mut self) -> State {
        one_of!(self.expression_literal_node())
    }
}
