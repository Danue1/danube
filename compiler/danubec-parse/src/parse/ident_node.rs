use crate::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::Context {
    pub fn ident_node(&mut self) -> State {
        guard!(self, IDENT_KEYWORD, IDENT_NODE);
        self.finish_node();
        State::Stop
    }
}
