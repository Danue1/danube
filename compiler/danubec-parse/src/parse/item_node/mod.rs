mod enum_item_node;
mod struct_item_node;

impl crate::Context {
    pub fn item_node(&mut self) -> crate::State {
        one_of!(self.struct_item_node(), self.enum_item_node())
    }
}
