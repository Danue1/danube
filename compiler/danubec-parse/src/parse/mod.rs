mod ident_node;
mod item_node;
mod path_node;
mod type_node;

use crate::context::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::context::Context {
    #[inline]
    pub fn ast(&mut self) {
        self.start_node(SyntaxKind::AST_NODE);
        self.skip_whitespace();
        while !self.is_empty() {
            if self.item_node() == State::Continue {
                self.unexpected_token();
            }
            self.skip_whitespace();
        }
        self.finish_node();
    }

    fn unexpected_token(&mut self) {
        self.start_node(SyntaxKind::ERROR);
        self.bump();
        self.finish_node();
    }
}
