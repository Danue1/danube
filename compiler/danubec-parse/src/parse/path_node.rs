use crate::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::Context {
    pub fn path_node(&mut self) -> State {
        guard!(self, IDENT_KEYWORD, PATH_NODE, ident_node);

        while !self.is_eof() {
            if expect!(self, SyntaxKind::COLON) {
                if expect!(self, SyntaxKind::COLON) {
                    self.skip_whitespace();
                    self.ident_node();
                } else {
                    self.unexpected_token();
                }
            } else {
                break;
            }
        }

        self.finish_node();
        State::Stop
    }
}
