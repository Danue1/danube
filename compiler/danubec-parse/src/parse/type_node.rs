use crate::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::Context {
    pub fn type_node(&mut self) -> State {
        let checkpoint = self.checkpoint();
        if self.unnamed_type_kind_node() == State::Continue
            && self.path_type_kind_node() == State::Continue
        {
            State::Continue
        } else {
            self.start_node_at(checkpoint, SyntaxKind::TYPE_NODE);
            self.finish_node();
            State::Stop
        }
    }

    // ()
    // (type1)
    // (type1, type2)
    pub fn unnamed_type_kind_node(&mut self) -> State {
        guard!(self, LEFT_PAREN, UNNAMED_TYPE_KIND_NODE);
        self.skip_whitespace();

        while !self.is_eof() {
            if expect!(self, SyntaxKind::RIGHT_PAREN) {
                break;
            }

            let checkpoint = self.checkpoint();
            if self.type_node() == State::Continue {
                self.unexpected_token();
            } else {
                self.start_node_at(checkpoint, SyntaxKind::UNNAMED_TYPE_KIND_NODE);
                self.finish_node();
            }
            self.skip_whitespace();
            if expect!(self, SyntaxKind::COMMA) {
                self.skip_whitespace();
            }
        }

        self.finish_node();
        State::Stop
    }

    // Foo
    // Foo::Bar
    pub fn path_type_kind_node(&mut self) -> State {
        let checkpoint = self.checkpoint();
        if self.path_node() == State::Continue {
            return State::Continue;
        }

        self.start_node_at(checkpoint, SyntaxKind::PATH_TYPE_KIND_NODE);
        self.finish_node();
        State::Stop
    }
}
