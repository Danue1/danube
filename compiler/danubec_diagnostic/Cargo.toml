[package]
name = "danubec_diagnostic"
version = "0.1.0"
edition = "2024"

[dependencies]
