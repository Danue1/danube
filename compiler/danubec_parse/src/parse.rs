use crate::{
    event::Event,
    grammar::{Context, krate},
};
use danubec_diagnostic::Diagnostic;
use danubec_lex::lex;
use danubec_syntax::{GreenNodeBuilder, SyntaxKind, SyntaxNode};

pub fn parse(source: &str, diagnostic: &mut Diagnostic) -> SyntaxNode {
    let mut parse = |tokens: &[_]| {
        let tokens: Vec<_> = tokens.iter().map(|&(kind, _)| kind).collect();
        let mut context = Context::new(&tokens, diagnostic);
        krate(&mut context);

        context.finish()
    };

    let tokens = lex(source);
    let events = parse(&tokens);

    build(&tokens, events)
}

fn build(mut tokens: &[(SyntaxKind, &str)], mut events: Vec<Event>) -> SyntaxNode {
    let mut builder = GreenNodeBuilder::new();

    macro_rules! advance {
        () => {
            if let Some(&(kind, text)) = tokens.get(0) {
                builder.token(kind.into(), text);
                tokens = &tokens[1..];
            }
        };
    }

    macro_rules! trivia {
        () => {
            loop {
                match tokens.get(0) {
                    Some((kind, _)) if kind.at_trivia() => advance!(),
                    _ => break,
                }
            }
        };
    }

    for index in 0..events.len() {
        match std::mem::replace(&mut events[index], Event::Placeholder) {
            event @ Event::Start { forward_parent, .. }
            | event @ Event::Expire { forward_parent } => {
                let mut forward_parents = match event {
                    Event::Start { kind, .. } => vec![kind],
                    _ => vec![],
                };
                let mut index = index;
                let mut forward_parent = forward_parent;

                while let Some(fp) = forward_parent {
                    index += fp;
                    match std::mem::replace(&mut events[index], Event::Placeholder) {
                        Event::Start {
                            kind,
                            forward_parent: fp,
                        } => {
                            forward_parents.push(kind);
                            forward_parent = fp;
                        }
                        Event::Expire { forward_parent: fp } => {
                            forward_parent = fp;
                        }
                        _ => {
                            //
                        }
                    }
                }

                for kind in forward_parents.into_iter().rev() {
                    builder.start_node(kind.into());
                }

                trivia!();
            }
            Event::Token => {
                advance!();
                trivia!();
            }
            Event::End => builder.finish_node(),
            Event::Placeholder => {
                //
            }
        }
    }

    SyntaxNode::new_root(builder.finish())
}
