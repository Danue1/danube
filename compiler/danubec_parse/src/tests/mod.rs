use crate::parse;
use danubec_diagnostic::Diagnostic;

#[test]
fn top_level_attribute() {
    let source = r#"
#![a("value")]
#![a(key)]
#![a(key1, key2)]
#![a(key = "value")]
#![a::b(key = "value")]"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn attribute() {
    let source = r#"
#[a("value")]
struct Foo;"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn zero_sized_struct() {
    let source = r#"
struct Foo;"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn named_struct() {
    let source = r#"
struct Foo { }
struct Bar {
    a: i32,
    b: String,
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn unnamed_struct() {
    let source = r#"
struct Foo();
struct Bar(());
struct Baz(i32, String);"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn mixed_enum() {
    let source = r#"
enum Foo {}
enum Bar {
    A,
    B { a: i32, b: String },
    C(i32, String),
    D = 42,
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn module() {
    let source = r#"
mod foo;
mod bar {
    struct Foo;
    enum Bar {
        A,
        B { a: i32, b: String },
        C(i32, String),
        D = 42,
    }
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn constant() {
    let source = r#"
const FOO: i32 = 42;
const BAR: str = "Hello, world!";"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn static_variable() {
    let source = r#"
static FOO: i32 = 42;
static BAR: str = "Hello, world!";"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn type_alias() {
    let source = r#"
type Foo = i32;
type Bar = (i32, String);"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn function() {
    let source = r#"
fn foo();
fn bar() {}
fn baz(a: i32, b: String) -> (i32, String) {
    (a, b)
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    dbg!(&diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn trait_definition() {
    let source = r#"
trait Foo { }
trait Bar {
    fn foo();
    const BAR: i32;
    type Baz;
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn anonymous_implement() {
    let source = r#"
impl Foo { }
impl Foo {
    fn foo() {}
    const BAR: i32 = 42;
    type Baz = i32;
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn targeted_implement() {
    let source = r#"
impl Foo for Bar { }
impl Foo for Bar {
    fn foo() {}
    const BAR: i32 = 42;
    type Baz = i32;
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn use_definition() {
    let source = r#"
use {};
use {{},*,foo,bar as baz,::{},::*,::foo,::bar as baz,};
use *;
use foo;
use foo as bar;
use foo::{};
use foo::{*,foo,bar as baz,{},::*,::foo,::bar as baz,::{}};

use ::{};
use ::{{},*,foo,bar as baz,::{},::*,::foo,::bar as baz,};
use ::*;
use ::foo;
use ::foo as bar;
use ::foo::{};
use ::foo::{*,foo,bar as baz,{},::*,::foo,::bar as baz,::{}};
"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    dbg!(&diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}
