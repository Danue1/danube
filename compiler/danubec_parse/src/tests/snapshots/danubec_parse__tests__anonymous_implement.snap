---
source: compiler/danubec_parse/src/tests/mod.rs
expression: node
---
KRATE_NODE@0..87
  NEW_LINE@0..1 "\n"
  DEFINITION_NODE@1..14
    IMPLEMENT_DEFINITION_NODE@1..14
      IMPL@1..5 "impl"
      WHITESPACE@5..6 " "
      PATH_TYPE_NODE@6..10
        PATH_NODE@6..10
          PATH_SEGMENT_IDENTIFIER_NODE@6..10
            IDENTIFIER_NODE@6..10
              IDENTIFIER_SEGMENT@6..10
                IDENTIFIER@6..9 "Foo"
                WHITESPACE@9..10 " "
      LEFT_BRACE@10..11 "{"
      WHITESPACE@11..12 " "
      RIGHT_BRACE@12..13 "}"
      NEW_LINE@13..14 "\n"
  DEFINITION_NODE@14..87
    IMPLEMENT_DEFINITION_NODE@14..87
      IMPL@14..18 "impl"
      WHITESPACE@18..19 " "
      PATH_TYPE_NODE@19..23
        PATH_NODE@19..23
          PATH_SEGMENT_IDENTIFIER_NODE@19..23
            IDENTIFIER_NODE@19..23
              IDENTIFIER_SEGMENT@19..23
                IDENTIFIER@19..22 "Foo"
                WHITESPACE@22..23 " "
      LEFT_BRACE@23..24 "{"
      NEW_LINE@24..25 "\n"
      WHITESPACE@25..29 "    "
      ASSOCIATED_DEFINITION_NODE@29..45
        FUNCTION_DEFINITION_NODE@29..45
          FN@29..31 "fn"
          WHITESPACE@31..32 " "
          IDENTIFIER_NODE@32..35
            IDENTIFIER_SEGMENT@32..35
              IDENTIFIER@32..35 "foo"
          LEFT_PAREN@35..36 "("
          RIGHT_PAREN@36..37 ")"
          WHITESPACE@37..38 " "
          FUNCTION_BODY_BLOCK_NODE@38..45
            BLOCK_EXPRESSION_NODE@38..45
              LEFT_BRACE@38..39 "{"
              RIGHT_BRACE@39..40 "}"
              NEW_LINE@40..41 "\n"
              WHITESPACE@41..45 "    "
      ASSOCIATED_DEFINITION_NODE@45..70
        CONSTANT_DEFINITION_NODE@45..70
          CONST@45..50 "const"
          WHITESPACE@50..51 " "
          IDENTIFIER_NODE@51..54
            IDENTIFIER_SEGMENT@51..54
              IDENTIFIER@51..54 "BAR"
          COLON@54..55 ":"
          WHITESPACE@55..56 " "
          PATH_TYPE_NODE@56..60
            PATH_NODE@56..60
              PATH_SEGMENT_IDENTIFIER_NODE@56..60
                IDENTIFIER_NODE@56..60
                  IDENTIFIER_SEGMENT@56..60
                    IDENTIFIER@56..59 "i32"
                    WHITESPACE@59..60 " "
          EQUAL@60..61 "="
          WHITESPACE@61..62 " "
          LITERAL_EXPRESSION_NODE@62..64
            INTEGER_LITERAL_NODE@62..64
              INTEGER_SEGMENT@62..64 "42"
          SEMICOLON@64..65 ";"
          NEW_LINE@65..66 "\n"
          WHITESPACE@66..70 "    "
      ASSOCIATED_DEFINITION_NODE@70..86
        TYPE_DEFINITION_NODE@70..86
          TYPE@70..74 "type"
          WHITESPACE@74..75 " "
          IDENTIFIER_NODE@75..79
            IDENTIFIER_SEGMENT@75..79
              IDENTIFIER@75..78 "Baz"
              WHITESPACE@78..79 " "
          EQUAL@79..80 "="
          WHITESPACE@80..81 " "
          PATH_TYPE_NODE@81..84
            PATH_NODE@81..84
              PATH_SEGMENT_IDENTIFIER_NODE@81..84
                IDENTIFIER_NODE@81..84
                  IDENTIFIER_SEGMENT@81..84
                    IDENTIFIER@81..84 "i32"
          SEMICOLON@84..85 ";"
          NEW_LINE@85..86 "\n"
      RIGHT_BRACE@86..87 "}"
